File: M3Tools/Dialogs/LoadScreenDialog.vb
Imports System.ComponentModel
Imports System.Windows.Forms

Public Class LoadScreenDialog
	Private Event DataChanging As EventHandler

	Private WithEvents LoadScreen As LoadingScreen

	Private ReadOnly Property LoadingScreen As LoadingScreen
		Get
			If LoadScreen Is Nothing OrElse LoadScreen.IsDisposed Then
				LoadScreen = New LoadingScreen
			End If

			Return LoadScreen
		End Get
	End Property

	Public Property LoadText As String
		Get
			Return LoadingScreen.LoadText
		End Get
		Set(value As String)
			LoadingScreen.LoadText = value
		End Set
	End Property

	Public Property Image As Drawing.Bitmap
		Get
			Return LoadingScreen.Image
		End Get
		Set(value As Drawing.Bitmap)
			LoadingScreen.Image = value
		End Set
	End Property

	<DefaultValue(False)>
	Public Property Closable As Boolean
		Get
			Return LoadingScreen.Closable
		End Get
		Set(value As Boolean)
			LoadingScreen.Closable = value
		End Set
	End Property

	Private Sub DialogClosed(sender As Object, e As EventArgs) Handles LoadScreen.DialogClosing
		'LoadScreen.Close()
		'__loadScreen = Nothing
	End Sub

	Public Function ShowDialog() As DialogResult
		LoadingScreen.Show()
	End Function

	Private Sub DialogDisposed(sender As Object, e As EventArgs) Handles Me.Disposed
		If LoadScreen IsNot Nothing Then
			LoadingScreen.Close()
			LoadingScreen.Dispose()
		End If
	End Sub

	Public Sub ShowError(message As String)
		Image = My.Resources.ErrorImage
		LoadText = message
	End Sub
End Class

File: M3Tools/Dialogs/LoadScreenDialog.cs
using System;
using System.ComponentModel;
using System.Windows.Forms;

namespace SPPBC.M3Tools
{

	public partial class LoadScreenDialog
	{
		private LoadingScreen _loadScreen;

		/// <summary>
		/// 
		/// </summary>
		public new void Dispose()
		{
			LoadingScreen.Close();
			_loadScreen = null;
			base.Dispose(true);
		}

		private LoadingScreen LoadingScreen
		{
			get
			{
				if (_loadScreen is null || _loadScreen.IsDisposed)
				{
					_loadScreen = new LoadingScreen();
				}

				return _loadScreen;
			}
		}

		/// <summary>
		/// The text to dispaly
		/// </summary>
		public string LoadText
		{
			get => LoadingScreen.LoadText;
			set => LoadingScreen.LoadText = value;
		}

		/// <summary>
		/// The image to display
		/// </summary>
		public System.Drawing.Bitmap Image
		{
			get => LoadingScreen.Image;
			set => LoadingScreen.Image = value;
		}

		/// <summary>
		/// Whether the loading screen is closable yet
		/// </summary>
		[DefaultValue(false)]
		public bool Closable
		{
			get => LoadingScreen.Closable;
			set => LoadingScreen.Closable = value;
		}

		/// <summary>
		/// Whether it should be the top most form in the application
		/// </summary>
		[DefaultValue(false)]
		public bool TopMost
		{
			get => LoadingScreen.TopMost;
			set => LoadingScreen.TopMost = value;
		}

		/// <summary>
		/// 
		/// </summary>
		public LoadScreenDialog()
		{
			InitializeComponent();
		}

		/// <summary>
		/// Show the loading screen
		/// </summary>
		/// <returns></returns>
		public DialogResult ShowDialog()
		{
			LoadingScreen.Show();
			return default;
		}

		private void DialogDisposed(object sender, EventArgs e)
		{
			if (LoadingScreen is not null)
			{
				LoadingScreen.Close();
				LoadingScreen.Dispose();
			}
		}

		/// <summary>
		/// Update the loading screen to show the error image and message
		/// </summary>
		/// <param name="message"></param>
		public void ShowError(string message)
		{
			Image = Properties.Resources.ErrorImage;
			LoadText = message;
		}
	}
}
